keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/ArtifactListener.kt

```diff
@@ -86,49 +82,65 @@ class ArtifactListener(
     if (event.controllerTriggered) {
       log.info("Fetching latest version of all registered artifacts...")
     }
-    syncDebArtifactVersions()
+    syncArtifactVersions()
   }
 
   /**
    * For each registered debian artifact, get the last version, and persist if it's newer than what we have.
    */
   // todo eb: should we fetch more than one version?
-  // todo eb: do this for docker images also
   @Scheduled(initialDelay = 60000, fixedDelayString = "\${keel.artifact-refresh.frequency:PT6H}")
-  fun syncDebArtifactVersions() =
+  fun syncArtifactVersions() =
     runBlocking {
-      artifactRepository.getAll(DEB).forEach { artifact ->
+      artifactRepository.getAll().forEach { artifact ->
         launch {
-          val lastRecordedVersion: String? = artifactRepository.versions(artifact).firstOrNull()
-          val latestVersion: String? = artifactService.getVersions(artifact.name).firstOrNull()
-          val latestAppVersion = "${artifact.name}-$latestVersion"
+          val lastRecordedVersion = getLatestStoredVersion(artifact)
+          val latestVersion = when (artifact) {
+            is DebianArtifact -> getLatestDeb(artifact)?.let { "${artifact.name}-$it" }
+            is DockerArtifact -> getLatestDockerTag(artifact)
+            else -> throw UnsupportedArtifactTypeException(artifact.type.value())
+          }
           if (latestVersion != null) {
             val hasNew = when {
               lastRecordedVersion == null -> true
-              latestAppVersion != lastRecordedVersion -> {
-                listOf(latestAppVersion, lastRecordedVersion).sortedWith(VERSION_COMPARATOR.reversed()).first() == latestAppVersion
+              latestVersion != lastRecordedVersion -> {
+                artifact.versioningStrategy.comparator.compare(lastRecordedVersion, latestVersion) > 0
               }
               else -> false
             }
 
             if (hasNew) {
               log.debug("Artifact {} has a missing version {}, persisting..", artifact, latestVersion)
-              val version = artifactService.getArtifact(artifact.name, latestVersion)
-              artifactRepository.store(artifact, latestAppVersion, debStatus(version))
+              val status = when (artifact.type) {
+                DEB -> debStatus(artifactService.getArtifact(artifact.name, latestVersion.removePrefix("${artifact.name}-")))
+                // todo eb: is there a better way to think of docker status?
```

Comment:
I'm not sure about status for Docker artifacts specifically, but as a side-comment, when reviewing this I was a bit confused with the fact that we name that type `ArtifactStatus`(which is somewhat open to interpretation, I think). Looking at the values of the enum, perhaps we could instead name it something like `ReleaseStatus` instead?

Going back to the question on Docker images, do we know if there's any life-cycle management of image releases like there is for Debian packages (meaning, there might be tags on the images with that release status, for instance)?

Reply:
Release status would definitely be better, and we can change that in another PR.

There isn't any built-in lifecycle management with docker images that I know of. I think at netflix we might add some sort of indication, but I don't want to support that now or worry about that now.